sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RegionScopedSessionContainer.java

```diff
@@ -0,0 +1,621 @@
+        this.hostName = hostName;
+        this.disableSessionCapturing = disableSessionCapturing;
+        this.globalEndpointManager = globalEndpointManager;
+        this.firstPreferredReadableRegionCached = new AtomicReference<>(StringUtils.EMPTY);
+        this.partitionKeyBasedBloomFilter = new PartitionKeyBasedBloomFilter();
+        this.regionScopedSessionCapturingOptionsAsString = stringifyConfig();
+    }
+
+    public RegionScopedSessionContainer(final String hostName, boolean disableSessionCapturing) {
+        this(hostName, disableSessionCapturing, null);
+    }
+
+    public RegionScopedSessionContainer(final String hostName) {
+        this(hostName, false, null);
+    }
+
+    public String getHostName() {
+        return this.hostName;
+    }
+
+    @Override
+    public void setDisableSessionCapturing(boolean value) {
+        this.disableSessionCapturing = value;
+    }
+
+    @Override
+    public boolean getDisableSessionCapturing() {
+        return this.disableSessionCapturing;
+    }
+
+    @Override
+    public String getSessionToken(String collectionLink) {
+
+        PathInfo pathInfo = new PathInfo(false, null, null, false);
+        PartitionScopedRegionLevelProgress partitionScopedRegionLevelProgress = null;
+
+        if (PathsHelper.tryParsePathSegments(collectionLink, pathInfo, null)) {
+            Long uniqueDocumentCollectionId = null;
+            if (pathInfo.isNameBased) {
+                String collectionName = PathsHelper.getCollectionPath(pathInfo.resourceIdOrFullName);
+                uniqueDocumentCollectionId = this.collectionNameToCollectionResourceId.get(collectionName);
+            } else {
+                ResourceId resourceId = ResourceId.parse(pathInfo.resourceIdOrFullName);
+                if (resourceId.getDocumentCollection() != 0) {
+                    uniqueDocumentCollectionId = resourceId.getUniqueDocumentCollectionId();
+                }
+            }
+
+            if (uniqueDocumentCollectionId != null) {
+                partitionScopedRegionLevelProgress = this.collectionResourceIdToPartitionScopedRegionLevelProgress.get(uniqueDocumentCollectionId);
+            }
+        }
+
+        if (partitionScopedRegionLevelProgress == null) {
+            return StringUtils.EMPTY;
+        }
+
+        return this.getCombinedSessionToken(partitionScopedRegionLevelProgress);
+    }
+
+    private PartitionScopedRegionLevelProgress getPartitionAndRegionScopedProgressForCollection(RxDocumentServiceRequest request) {
+        return getPartitionAndRegionScopedProgressForCollection(request.getIsNameBased(), request.getResourceId(), request.getResourceAddress());
+    }
+
+    // fetch progress / session information for a given collection
+    private PartitionScopedRegionLevelProgress getPartitionAndRegionScopedProgressForCollection(boolean isNameBased, String rId, String resourceAddress) {
```

Comment:
seems can be replaced by `getCollectionRidToPartitionScopedRegionLevelProgress`?

Reply:
Fixed in next iteration.